ladybird: unstable-2023-01-17 -> unstable-2024-03-16

ADKaster opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Update to current master.

- Pre-Download required Unicode, Common Locale Database, Timezone database and Root CA certificates, and ICC profiles to build with the new option -DENABLE_NETWORK_DOWNLOADS=OFF
- Copy required downloaded files to the expected SERENITY_CACHE_DIR layout
- Remove workarounds for resolved upstream issues.
- Add new missing qtmultimedia package
- Remove unused unzip package
- Point homepage at https://ladybird.dev
- Set mainProgram to "ladybird"
- Enable the macOS bundle on Darwin platforms


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

cc @emilytrau @networkException @fgaz

ADKaster commented:
Fixes #212792

networkException commented on pkgs/applications/networking/browsers/ladybird/default.nix:
```diff
@@ -1,71 +1,117 @@
 }:
 
+let
+  cldr_version = "43.1.0";
+  ucd_version = "15.0.0";
+  emoji_txt_version = "15.0";
+  tzdb_version = "2023c";
+  cacert_version = "2023-05-30";
+  cldr = fetchzip {
+    url = "https://github.com/unicode-org/cldr-json/releases/download/${cldr_version}/cldr-${cldr_version}-json-modern.zip";
+    stripRoot = false;
+    hash = "sha256-+PCYYUWwiDyTfDvSxBAbVRpA7YkuQGKr3ZPlnc9PnKk=";
+  };
+  ucd = fetchzip {
+    url = "https://www.unicode.org/Public/${ucd_version}/ucd/UCD.zip";
+    stripRoot = false;
+    hash = "sha256-jj6bX46VcnH7vpc9GwM9gArG+hSPbOGL6E4SaVd0s60=";
+  };
+  emoji_txt = fetchurl {
+    url = "https://www.unicode.org/Public/emoji/${emoji_txt_version}/emoji-test.txt";
+    hash = "sha256-hEXyOsg4jglr4Z0CYuFPzv+Fb/Ugk/I1bciUhfGoU9s=";
+  };
+  tzdb = fetchzip {
+    url = "https://data.iana.org/time-zones/releases/tzdata${tzdb_version}.tar.gz";
+    stripRoot = false;
+    hash = "sha256-G/IC0EHQcuXUTr2G/6Kn1giC0d1WC9hHIE2fQI26rlM=";
+  };
+  cacert = fetchurl {
+    url = "https://curl.se/ca/cacert-${cacert_version}.pem";
+    hash = "sha256-X63K6QqkrgQRUPji0mw32YBSLNtJ+SP8HhteuNdOca0=";
+  };
```
In general these should probably be dependencies on other packages wherever possible

ADKaster replied:
hm ok. 

looks like we could use `unicode-character-database` for the UCD download, the versions match (atm). But, the build step would still need to have the `version.txt` file in the location of the database at build time, and not at runtime.

there's a `tzdata` package as well, but I'm not sure whether it actually installs the raw extracted files, or simply compiles them into a more distro-usable format.

The `cldr-annotations` package doesn't use the JSON download of the CLDR release, and seems to only grab the "common" files, whereas we use more than just those.

we can definitely use the cacert package though

qbit replied:
I wonder if you could use the `.src` (or `.srcs`?) of tzdata?

networkException replied:
I would assume best effort with whatever compatible packages exist currently is fine, but a committer could have a different opinion. In that case I'll happily help get the other things available as individual packages

ADKaster replied:
I was able to eliminate all of the manual fetches except for the CLDR json by using existing packages.

fgaz commented:
> Alternatively, we could use https://ladybird.dev

Is that the official website? I can't find any links to it in the repo or at serenityos.org

networkException commented:
Yes its official

qbit commented:
Result of `nixpkgs-review pr 248456` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ladybird</li>
  </ul>
</details>

fgaz commented:
> Yes its official, though only really documented on Discord I guess.

Confirmed. Please use https://ladybird.dev then

fgaz commented on pkgs/applications/networking/browsers/ladybird/default.nix:
```diff
@@ -1,71 +1,117 @@
+    mkdir Caches/CACERT
+    cp -r ${cacert} Caches/CACERT/cacert-${cacert_version}.pem
+
+    chmod +w Caches/UCD
+    chmod +w Caches/CLDR
+    chmod +w Caches/TZDB
+    chmod +w Caches/CACERT
+
+    cp ${emoji_txt} Caches/UCD/emoji-test.txt
+
+    echo -n ${cldr_version} > Caches/CLDR/version.txt
+    echo -n ${ucd_version} > Caches/UCD/version.txt
+    echo -n ${tzdb_version} > Caches/TZDB/version.txt
+    echo -n ${cacert_version} > Caches/CACERT/version.txt
+
+    chmod -w Caches/UCD
+    chmod -w Caches/CLDR
+    chmod -w Caches/TZDB
+    chmod -w Caches/CACERT
```
This can be done with `postFetch` in `fetchzip` (unless we end up doing separate packages as was suggested in the other thread)

ADKaster replied:
I did this for the CLDR json. I think the current setup with preConfigure looks "slightly" better than the way I had it before, but still not super great. Since I pulled the rest of the files from existing packages it still looks a bit gnarly.

fgaz reviewed (commented):
Does not appear to build on darwin, please mark it as [broken](https://nixos.org/manual/nixpkgs/stable/index.html#var-meta-broken) (or fix it)

ADKaster commented:
> Does not appear to build on darwin, please mark it as [broken](https://nixos.org/manual/nixpkgs/stable/index.html#var-meta-broken) (or fix it)

I tried testing it on my apple silicon MacBook and I couldn't get past trying to get the build to use clang-11. No matter what hax I tried it kept wanting to use `clang-11`, which is far too old to compile the project (it actually crashes the compiler). I was able to force it to use one of the llvmPackages_16 derivations, but then the compile couldn't find the Apple SDK $SYSROOT or something, `<assert.h>` was not found.

Either way, adding 
```
    substituteInPlace CMakeLists.txt \
      --replace "if (APPLE)" "if (FALSE)"
```
let the CMake configure, so I'll add that and see if it fixes the CI? If that fails I'll mark it as broken :\

ADKaster commented:
Well that's super annoying. Why does nix pass an absolute path for both CMAKE_INSTALL_BINDIR  and CMAKE_INSTALL_LIBDIR? Aren't those both supposed to be relative to CMAKE_INSTALL_PREFIX? I pushed a fix to serenity upstream that I figured would fix the macOS builds but ended up breaking the Linux ones :/

ADKaster commented:
All right, x86_64 Darwin is now working, but aarch64-darwin CI is seeing the same "using clang-11 even though we told it to use a different compiler" issue I was seeing locally. So not sure if I should mark that as unsupported, report an issue/discourse thread, or accept it as expected?

fgaz commented:
Perhaps try using `darwin.apple_sdk_11_0.callPackage` instead of `callPackage` (last bullet at https://nixos.org/manual/nixpkgs/stable/index.html#sec-darwin )

fgaz commented on pkgs/applications/networking/browsers/ladybird/default.nix:
```diff
@@ -1,71 +1,109 @@
+    cp -r ${unicode-character-database}/share/unicode build/Caches/UCD
+    chmod +w build/Caches/UCD
+    cp ${unicode-emoji}/share/unicode/emoji/emoji-test.txt build/Caches/UCD
+    echo -n ${ucd_version} > build/Caches/UCD/version.txt
```
Can we get the versions from the packages? For example:

```suggestion
    echo -n ${unicode-emoji.version} > build/Caches/UCD/version.txt
```

ADKaster replied:
If the version in the cache dir doesn't exactly match the version in our CMake download helpers, it will reject the fileb and try to download from the Internet. We would either need to make sure they always line up when updating the package, or lie to the Meta/CMake files with more patches/replacements.

ADKaster replied:
I went ahead and did this for tzdata and unicode-character-database (the emoji-test.txt version is 15.0, not 15.0.0), and left some notes to future maintainers about the way the CMake files are particular about the exact versions matching.

ADKaster commented:
> Perhaps try using darwin.apple_sdk_11_0.callPackage instead of callPackage (last bullet at https://nixos.org/manual/nixpkgs/stable/index.html#sec-darwin )

all-packages.nix currently has this:
```
 ladybird = qt6Packages.callPackage ../applications/networking/browsers/ladybird {
    stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.llvmPackages_14.stdenv else stdenv;
  };
```

I did a check of `env | grep PATH` with the current setup, and saw that `clang-wrapper-11` was *before* `clang-wrapper-14` in the path on my machine.

Which seems... guaranteed not to work?
```
PATH=/nix/store/32w82ry9cb7bsxdx324wjnbdgh989dqk-cmake-3.26.4/bin
/nix/store/69dxkkb3vwx73n332fldjvll393cpydp-ps-adv_cmds-119/bin
/nix/store/jcd8f778j0ckdmp50h9wn8iakkclf2cd-ninja-1.11.1/bin
/nix/store/7dkbcs6vgwbkb6rafh03j9j0bn8qpki0-clang-wrapper-11.1.0/bin
/nix/store/78g551gc2wkzr06ar571fgvj6qn5gi14-clang-11.1.0/bin
/nix/store/r30ljn3w4317s0xhf8s6wf3c3kx5cfiv-coreutils-9.3/bin
/nix/store/dnvdnqc5cdx499j2y0pyqlsdvd868y08-cctools-binutils-darwin-wrapper-11.1.0-973.0.1/bin
/nix/store/6j9rsw4086h5g8db7anxy6bbrsv4zqz7-cctools-binutils-darwin-11.1.0-973.0.1/bin
/nix/store/3h3nfz7qbf8f9j2acvsdaqs59dd0pkkp-lndir-1.0.4/bin
/nix/store/fgap77lk5a7q9xjan91cf135bvymvb4m-clang-wrapper-14.0.6/bin
/nix/store/n8zdf0a7zz7jk4md936j97g5gykhl5sc-clang-14.0.6/bin
```

If I change it to this (with the required `with qt6Packages;` additions to nativeBuildInputs and buildInputs):

```
  ladybird = darwin.apple_sdk_11_0.callPackage ../applications/networking/browsers/ladybird {
    stdenv = if stdenv.isDarwin then darwin.apple_sdk_11_0.llvmPackages_16.stdenv else stdenv;
  };
```

Then I see the same, but with clang-wrapper-16 at the position that clang-wrapper-14 was at:
```
PATH=/nix/store/32w82ry9cb7bsxdx324wjnbdgh989dqk-cmake-3.26.4/bin
/nix/store/69dxkkb3vwx73n332fldjvll393cpydp-ps-adv_cmds-119/bin
/nix/store/jcd8f778j0ckdmp50h9wn8iakkclf2cd-ninja-1.11.1/bin
/nix/store/7dkbcs6vgwbkb6rafh03j9j0bn8qpki0-clang-wrapper-11.1.0/bin
/nix/store/78g551gc2wkzr06ar571fgvj6qn5gi14-clang-11.1.0/bin
/nix/store/r30ljn3w4317s0xhf8s6wf3c3kx5cfiv-coreutils-9.3/bin
/nix/store/dnvdnqc5cdx499j2y0pyqlsdvd868y08-cctools-binutils-darwin-wrapper-11.1.0-973.0.1/bin
/nix/store/6j9rsw4086h5g8db7anxy6bbrsv4zqz7-cctools-binutils-darwin-11.1.0-973.0.1/bin
/nix/store/3h3nfz7qbf8f9j2acvsdaqs59dd0pkkp-lndir-1.0.4/bin
/nix/store/3dl7pdy82jdzbxixgbvvdl89hq5mh8dz-clang-wrapper-16.0.6/bin
/nix/store/c52xq08a10zqkazj4q0a3y4md672505g-clang-16.0.6/bin
```
Is this supposed to work?

ADKaster commented on pkgs/applications/networking/browsers/ladybird/default.nix:
```diff
@@ -1,62 +1,118 @@
+    "-DSERENITY_CACHE_DIR=Caches"
+    "-DENABLE_NETWORK_DOWNLOADS=OFF"
+    "-DENABLE_COMMONMARK_SPEC_DOWNLOAD=OFF"
+  ] ++ lib.optionals stdenv.isDarwin [
+    # Force using newer clang for Apple Silicon platforms,
+    # as just using the stdenv still gets us clang-11 by default.
+    "-DCMAKE_CXX_COMPILER=${llvmPackages.clang}/bin/clang++"
+    "-DCMAKE_C_COMPILER=${llvmPackages.clang}/bin/clang"
   ];
```
I updated the callPackages as recommend in the comments, but I was still unable to build the package on my Apple Silicon MacBook without explicitly requesting the clang/clang++ binaries from the `darwin.apple_sdk_11_0.llvmPackages_16` package. The stdenv just kept making bare `clang` and `clang++` point to the `clang-wrapper-11` package instead of `clang-wrapper-16` like I was expecting.

ADKaster committed:
ladybird: unstable-2023-01-17 -> 0-unstable-2024-03-16

ADKaster commented:
@fgaz  I've updated the ladybird PR here to the latest current rev and tested it on both macOS Sonoma (apple silicon) and Ubuntu 22.04 (x86_64).

There were a few extra upstream changes since the last time I looked at this package:

- The app name was changed to `Ladybird` with a capital L
- The macOS app bundle should be fully self-contained now
- Most internal helper process executables were moved to $PREFIX/libexec. The way that Ladybird/CMakeLists.txt reads CMAKE_INSTALL_LIBEXECDIR to embed in the binary really really wants the path to be relative to $CMAKE_INSTALL_PREFIX rather than an absolute path. I'm assuming that nixpkgs would rather we patch the CMakeLists.txt than pass a relative path with -DCMAKE_INSTALL_LIBEXECDIR=libexec?
- A few more downloaded files were added, including another Unicode file (IDNAMappingTable.txt), the public suffix list (upstream cmake just downloads `master`, but I pinned it), and a standard ICC profile from Adobe.
- The macOS port by default uses a custom AppKit/Cocoa UI instead of Qt. More patches are needed to properly support both UI toolkits in the nix file.
- OpenGL is used for graphics-accelerated painting, and WebGL features. I could not get this to work properly on GNU/Linux (Ubuntu), so I disabled the feature on Linux.
- There's CMake feature detection on non-macOS to use pulse audio if available. If pulse audio is not available, it falls back to Qt multimedia.

2xsaiko commented:
> * Most internal helper process executables were moved to $PREFIX/libexec. The way that Ladybird/CMakeLists.txt reads CMAKE_INSTALL_LIBEXECDIR to embed in the binary really really wants the path to be relative to $CMAKE_INSTALL_PREFIX rather than an absolute path. I'm assuming that nixpkgs would rather we patch the CMakeLists.txt than pass a relative path with -DCMAKE_INSTALL_LIBEXECDIR=libexec?

If you add something like this to make the absolute path relative into the upstream CMakeLists.txt, then it should just work for both cases: https://git.dblsaiko.net/nucom/tree/CMakeLists.txt#n32

fgaz commented:
Thanks, I'll have a look tomorrow. Merging this is really long overdue :sweat_smile: 

`-DCMAKE_INSTALL_LIBEXECDIR=libexec` is also fine if you can't use an absolute path, we have many packages doing that

fgaz reviewed (approved):
Looks good, thank you for all the work!
